ext/cosmos/ext/structure/structure.c

```diff
@@ -902,13 +902,20 @@ static VALUE binary_accessor_write(VALUE self, VALUE value, VALUE param_bit_offs
       if (given_bit_size <= 0) {
         end_bytes = -(given_bit_size / 8);
         old_upper_bound = buffer_length - 1 - end_bytes;
+        /* If the buffer is not already big enough to hold the non-variable sized items
+             that is a buffer error.  old_upper_bound can be -1 if the string is exactly the size of the
+             end_bytes which is ok */
+        if (old_upper_bound < -1) {
+          rb_funcall(self, id_method_raise_buffer_error, 5, symbol_write, param_buffer, param_data_type, param_bit_offset, param_bit_size);
+        }
+
         if (old_upper_bound < lower_bound) {
           /* String was completely empty */
           if (end_bytes > 0) {
             /* Preserve bytes at end of buffer */
             rb_str_concat(param_buffer, rb_str_times(ZERO_STRING, INT2FIX(value_length)));
             buffer = (unsigned char*) RSTRING_PTR(param_buffer);
-            memmove((buffer + lower_bound + value_length), (buffer + lower_bound), value_length);
+            memmove((buffer + lower_bound + value_length), (buffer + lower_bound), end_bytes);
```

Comment:
Hold off merging.  It still seems like lower_bound needs to be accounted for somewhere.